bot/cogs/doc.py

```diff
@@ -0,0 +1,207 @@
+import functools
+import logging
+import re
+import sys
+from collections import OrderedDict
+from typing import Optional, Tuple
+
+import aiohttp
+import discord
+from bs4 import BeautifulSoup
+from discord.ext import commands
+from markdownify import MarkdownConverter
+from sphinx.ext import intersphinx
+
+log = logging.getLogger(__name__)
+
+
+BASE_URLS = {
+    'aiohttp': "https://aiohttp.readthedocs.io/en/stable/",
+    'discord': "https://discordpy.readthedocs.io/en/rewrite/",
+    'django': "https://docs.djangoproject.com/en/dev/",
+    'stdlib': "https://docs.python.org/%d.%d/" % sys.version_info[:2]
+}
+
+INTERSPHINX_INVENTORIES = {
+    'aiohttp': "https://aiohttp.readthedocs.io/en/stable/objects.inv",
+    'discord': "https://discordpy.readthedocs.io/en/rewrite/objects.inv",
+    'django': "https://docs.djangoproject.com/en/dev/_objects/",
+    'stdlib': "https://docs.python.org/%d.%d/objects.inv" % sys.version_info[:2]
+}
+
+UNWANTED_SIGNATURE_SYMBOLS = ('[source]', '¶')
+WHITESPACE_AFTER_NEWLINES_RE = re.compile(r"(?<=\n\n)( +)")
+
+
+def async_cache(max_size=128, arg_offset=0):
+    """
+    LRU cache implementation for coroutines.
+
+    :param max_size:
+    Specifies the maximum size the cache should have.
+    Once it exceeds the maximum size, keys are deleted in FIFO order.
+    :param arg_offset:
+    The offset that should be applied to the coroutine's arguments
+    when creating the cache key. Defaults to `0`.
+    """
+
+    cache = OrderedDict()
```

Comment:
Dicts are ordered by default. To remove a key (R58) use `del cache[list(cache.keys())[0]]` instead to save memory.

Reply:
> The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon (this may change in the future, but it is desired to have this new dict implementation in the language for a few releases before changing the language spec to mandate order-preserving semantics for all current and future Python implementations; this also helps preserve backwards-compatibility with older versions of the language where random iteration order is still in effect, e.g. Python 3.5).
from [What's New in Python 3.6](https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-compactdict) - I also thought it would be more intuitive to use this here.

Reply:
`del cache[list(cache.keys())[0]]`  looks horrible. OrderedDicts are fine. dicts are not guaranteed to retain insertion order until 3.7, and even then there are compelling reasons to just use OrderedDict anyway because of readability. The "save memory" argument has no place in a project of this nature. Code readability trumps marginal memory saving every time.